### PR TITLE
feat(tracer): add Python 3.11 support

### PR DESCRIPTION
**Pull request by Yun-Kim**
## Description
Python 3.11 has made some new changes to their C internal API that we need to address in a couple of places, mostly in our profiler. I've moved the profiler changes into #4343 and changed this PR to focus on the tracer and integrations supporting Python 3.11. Note that some of the profiler changes were kept in this PR to ensure that the profiler can compile with Python 3.11 per our setup/install rules.

The major tracer-related fix is:
- Python 3.11 moved `_PyFloat_Pack8()` into their internal C API and replaced it with `PyFloat_Pack8()` to their public C API ([Link to CPython issue](https://github.com/python/cpython/issues/91062)). We use it once in `ddtrace.internal.pack_template.h` where I've added a Python version check to use the correct corresponding function.

Note: grpcio with pytest-asyncio is causing Python segmentation faults on the `test_unary_exception` and `test_unary_cancellation` test cases, and from the traceback it doesn't look like dd-trace-py is involved. You can reproduce this using this [gist](https://gist.github.com/Yun-Kim/e4a8f3cef807caa3d42cd9c1714614de) which does not involve dd-trace-py. As a result I am skipping the grpcio_asyncio tests with Python 3.11.

## Checklist
- [x] Title must conform to [conventional commit](https://github.com/conventional-changelog/commitlint/tree/master/%40commitlint/config-conventional).
- [x] Add additional sections for `feat` and `fix` pull requests.
- [ ] Ensure tests are passing for affected code.
- [ ] [Library documentation](https://github.com/DataDog/dd-trace-py/tree/1.x/docs) and/or [Datadog's documentation site](https://github.com/DataDog/documentation/) is updated. Link to doc PR in description.

## Reviewer Checklist
- [x] Title is accurate.
- [x] Description motivates each change.
- [x] No unnecessary changes were introduced in this PR.
- [x] PR cannot be broken up into smaller PRs.
- [x] Avoid breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Tests provided or description of manual testing performed is included in the code or PR.
- [x] Release note has been added for fixes and features, or else `changelog/no-changelog` label added.
- [x] All relevant GitHub issues are correctly linked.
- [x] Backports are identified and tagged with Mergifyio.
- [ ] Add to milestone.

### CODE REVIEW

File: docker/.python-version
Comment (brettlangdon) [inline docker/.python-version:8]: do we care here? `-dev` should get us latest rc no?

also this change needs to happen in it's own PR since CircleCI uses the _published_ image, which won't get published until this change is merged into `1.x` branch

File: setup.py
Comment (brettlangdon) [inline setup.py:315]: good luck

File: docker/.python-version
Comment (Yun-Kim) [reply docker/.python-version]: Sounds good, does this mean that I can just revert this back to `3.11-dev` and wait until 3.11 is officially released?

Comment (brettlangdon) [reply docker/.python-version]: yes, it should be ok to revert it

File: ddtrace/internal/pack_template.h
Comment (P403n1x87) [inline ddtrace/internal/pack_template.h:562]:
```suggestion
    #if PY_VERSION_HEX < 0x030B0000
```

File: ddtrace/internal/pack_template.h
Comment (P403n1x87) [inline ddtrace/internal/pack_template.h:561]:
```suggestion
    // Python 3.11 introduced PyFloat_Pack8() to the public C API and moved _PyFloat_Pack8() to the internal C API
```
?

File: releasenotes/notes/python311-a7077c0e461622d2.yaml
Comment (P403n1x87) [inline releasenotes/notes/python311-a7077c0e461622d2.yaml:4]: newline

**Comment by P403n1x87** [comment]
Just a note that the `bytecode` module doesn't support Python 3.11 yet, so the `debugger` and `internal` test suites won't pass.

**Comment by Yun-Kim** [comment]
The `bytecode` [library](https://github.com/MatthieuDartiailh/bytecode) currently does not support Python 3.11, and is used by the `internal, debugger, graphene, graphql` components/integrations. Until then, I'll exclude those test suites from running Python 3.11.

**Comment by jd** [comment]
I've replied to @Yun-Kim already, just leaving a trace here: there's an infinite  loop in the memory profiler with Python 3.11 which I was able to identify. I'll work on a fix.

**Comment by jd** [comment]
@Yun-Kim I've a fix for the memory part, but it seems there's another issue with the changes on the stack collector
```
Exception in thread ddtrace.profiling.collector.stack:StackCollector:
Traceback (most recent call last):
  File "/root/.pyenv/versions/3.11-dev/lib/python3.11/threading.py", line 1038, in _bootstrap_inner
    self.run()
  File "/root/project/ddtrace/internal/periodic.py", line 73, in run
    self._target()
  File "/root/project/ddtrace/profiling/collector/__init__.py", line 42, in periodic
    for events in self.collect():
                  ^^^^^^^^^^^^^^
  File "ddtrace/profiling/collector/stack.pyx", line 500, in ddtrace.profiling.collector.stack.StackCollector.collect
    all_events = stack_collect(

  File "ddtrace/profiling/collector/stack.pyx", line 382, in ddtrace.profiling.collector.stack.stack_collect
    frames, nframes = _traceback.traceback_to_frames(exc_traceback, max_nframes)

  File "ddtrace/profiling/collector/_traceback.pyx", line 23, in ddtrace.profiling.collector._traceback.traceback_to_frames
    cpdef traceback_to_frames(traceback, max_nframes):

  File "ddtrace/profiling/collector/_traceback.pyx", line 35, in ddtrace.profiling.collector._traceback.traceback_to_frames
    frame = tb.tb_frame

AttributeError: type object 'type' has no attribute 'tb_frame'
```

Not sure what it is yet. The code looks ok to me, but there might be something we missed.

**Comment by jd** [comment]
memalloc should be fixed with: https://github.com/DataDog/dd-trace-py/pull/4304

**Comment by Yun-Kim** [comment]
> @Yun-Kim I've a fix for the memory part, but it seems there's another issue with the changes on the stack collector
> 
> ```
> Exception in thread ddtrace.profiling.collector.stack:StackCollector:
> Traceback (most recent call last):
>   File "/root/.pyenv/versions/3.11-dev/lib/python3.11/threading.py", line 1038, in _bootstrap_inner
>     self.run()
>   File "/root/project/ddtrace/internal/periodic.py", line 73, in run
>     self._target()
>   File "/root/project/ddtrace/profiling/collector/__init__.py", line 42, in periodic
>     for events in self.collect():
>                   ^^^^^^^^^^^^^^
>   File "ddtrace/profiling/collector/stack.pyx", line 500, in ddtrace.profiling.collector.stack.StackCollector.collect
>     all_events = stack_collect(
> 
>   File "ddtrace/profiling/collector/stack.pyx", line 382, in ddtrace.profiling.collector.stack.stack_collect
>     frames, nframes = _traceback.traceback_to_frames(exc_traceback, max_nframes)
> 
>   File "ddtrace/profiling/collector/_traceback.pyx", line 23, in ddtrace.profiling.collector._traceback.traceback_to_frames
>     cpdef traceback_to_frames(traceback, max_nframes):
> 
>   File "ddtrace/profiling/collector/_traceback.pyx", line 35, in ddtrace.profiling.collector._traceback.traceback_to_frames
>     frame = tb.tb_frame
> 
> AttributeError: type object 'type' has no attribute 'tb_frame'
> ```
> 
> Not sure what it is yet. The code looks ok to me, but there might be something we missed.

@jd Interesting, from my understanding I've merely replaced the exception type & traceback with utility functions instead of accessing them directly, but it appears that the traceback object is actually a `type` object? Do you think there might've been some implicit switching of the exception tuple's ordering?

**Comment by Yun-Kim** [comment]
> memalloc should be fixed with: #4304

@jd Thank you for the fix! 🎉

File: riotfile.py
Comment (Kyle-Verhoog) [inline riotfile.py:92]: Let's propose this change in a new PR, there isn't a 3.11 reason to do it

File: riotfile.py
Comment (Kyle-Verhoog) [inline riotfile.py:251]: It looks like this change is a no-op, leave it out?

File: riotfile.py
Comment (Kyle-Verhoog) [inline riotfile.py:260]:
```suggestion
                # Riot venvs break with Py 3.11 importlib, specifically with hypothesis (test_http.py). We run the test_http.py tests with tox in CI.
```

can we also move this line down to above the command line

File: riotfile.py
Comment (Kyle-Verhoog) [inline riotfile.py:262]:
```suggestion
```

File: riotfile.py
Comment (Kyle-Verhoog) [inline riotfile.py:311]: This means that we're not running any of the other internal tests against 3.11 due to bytecode.

There are a lot of tests in internal which are not dependent on `bytecode`. 


We could do this hack for now in `tests/internal/__init__.py`:

```python
import sys
import mock

if sys.version_info >= (3, 11, 0):
    sys.modules["bytecode"] = mock.MagicMock()
    import bytecode
```

not great but at least let's us run most of our tests still

File: riotfile.py
Comment (Kyle-Verhoog) [inline riotfile.py:336]: Worth having a comment on why we split the suite (cause of gevent).

Also, since we have a bunch of gevent work-arounds for 3.11 can we add a common marker so that when gevent adds support for 3.11 we can easily find all the places we need to update? Maybe something like `FIXME[gevent-3.11]: ...`?

File: riotfile.py
Comment (Kyle-Verhoog) [inline riotfile.py:351]: Comment/FIXME about how debugger doesn't have 3.11 support

File: riotfile.py
Comment (Kyle-Verhoog) [inline riotfile.py:612]: The django docs linked below actually say that it's 4.2 that adds support for 3.11.

File: riotfile.py
Comment (Kyle-Verhoog) [inline riotfile.py:646]: Maybe revert these changes? They don't seem relevant to 3.11 (unless the suites actually fail).

File: riotfile.py
Comment (Kyle-Verhoog) [inline riotfile.py:1401]: Can we add a comment why we impose a starlette version here? (It overrides the dependency installed by fastapi)

File: riotfile.py
Comment (Kyle-Verhoog) [inline riotfile.py:1614]: Can we do a `FIXME[bytecode-3.11]`

File: tox.ini
Comment (Kyle-Verhoog) [inline tox.ini:24]: Worth having a comment here about why this suite exists (the work-around for riot issue). Maybe `FIXME[riot-3.11]`

File: tests/internal/test_forksafe.py
Comment (Kyle-Verhoog) [inline tests/internal/test_forksafe.py:291]: @P403n1x87 I took a quick look but couldn't figure out why the skips weren't working. My hypothesis is that the hook we use in pytest for the subprocess tests runs before the hook that checks for skipping (but haven't been able to confirm)

**Review by Kyle-Verhoog — DISMISSED** [review]

File: tests/internal/test_forksafe.py
Comment (P403n1x87) [reply tests/internal/test_forksafe.py:291]: That's probably right. I don't know if there's an easy way to control that. Perhaps we could add a `skip` argument to handle this

File: tox.ini
Comment (majorgreys) [inline tox.ini:74]: Why is 3.10 being removed?

File: releasenotes/notes/python311-tracer-a7077c0e461622d2.yaml
Comment (Yun-Kim) [inline releasenotes/notes/python311-tracer-a7077c0e461622d2.yaml:4]: add verbose prelude mentioning initial support for tracer, profiler and dynamic instrumentation is not yet compatible.

File: riotfile.py
Comment (Yun-Kim) [inline riotfile.py:263]: Might be worth adding the repro into a gist for future. attach into riot issue

Comment (Yun-Kim) [reply riotfile.py:263]: link riot issue

**Comment by majorgreys** [comment]
While this is not full support with wheels, we should update https://ddtrace.readthedocs.io/en/stable/versioning.html#supported-runtimes.

File: tests/contrib/grpc_aio/test_grpc_aio.py
Comment (Yun-Kim) [inline tests/contrib/grpc_aio/test_grpc_aio.py:399]: Add/link upstream issue

Comment (Yun-Kim) [reply tests/contrib/grpc_aio/test_grpc_aio.py:399]: Document in supported libraries with asterisk (explain why xyz library doesn't support 3.11)

File: tox.ini
Comment (Yun-Kim) [reply tox.ini:74]: Add comment explaining gevent compatibility issue with opentracer

File: pyproject.toml
Comment (Kyle-Verhoog) [inline pyproject.toml:20]: why did we remove py27?

File: releasenotes/notes/python311-tracer-a7077c0e461622d2.yaml
Comment (Kyle-Verhoog) [inline releasenotes/notes/python311-tracer-a7077c0e461622d2.yaml:4]:
```suggestion
    Initial library support has been added for Python 3.11. Profiling and Dynamic Instrumentation are
```

File: pyproject.toml
Comment (Yun-Kim) [reply pyproject.toml]: Note: we're running our formatting pre-check using `black==21.4b4` which doesn't support python 3.10 or python 3.11. I'll revert the changes instead until we upgrade our black version.

**Review by Kyle-Verhoog — APPROVED** [review]

File: ddtrace/profiling/collector/stack.pyx
Comment (Yun-Kim) [reply ddtrace/profiling/collector/stack.pyx:252]: This code shouldn't differ functionally from the previous version, as it's been stated by CPython maintainers that `exc_info.exc_type` and `exc_info.exc_traceback` are redundant with `exc_info.exc_value` and can be inferred from `exc_info.exc_value`. Since we're doing a null check here on the exception object and its fields, this check is effectively equivalent to the previous version.

However you're right, this change isn't necessary pre-3.11, but can be fixed in #4343 since the goal of this PR is to unblock the tracer from the profiler and ensure the profiler can compile.

**Review by Kyle-Verhoog — APPROVED** [review]

**Review by Kyle-Verhoog — APPROVED** [review]

File: ddtrace/debugging/_debugger.py
Comment (majorgreys) [inline ddtrace/debugging/_debugger.py:156]: What would a user see if they tried using the profiler after installing 3.11?

File: releasenotes/notes/python311-tracer-a7077c0e461622d2.yaml
Comment (majorgreys) [inline releasenotes/notes/python311-tracer-a7077c0e461622d2.yaml:4]:
```suggestion
    Initial library support has been added for Python 3.11. Continuous Profiler and Dynamic Instrumentation are not yet compatible and must be disabled in order to use the library with Python 3.11. Support for Continuous Profiler and Dynamic Instrumentation will be added in a future release.
```

File: releasenotes/notes/python311-tracer-a7077c0e461622d2.yaml
Comment (majorgreys) [inline releasenotes/notes/python311-tracer-a7077c0e461622d2.yaml:8]:
```suggestion
features:
  - |
    tracer: added support for Python 3.11.
upgrade:
  - |
    Python 3.11: Continuous Profiler and Dynamic Instrumentation must be disabled as they are not currently supported.
```

**Review by majorgreys — DISMISSED** [review]

**Review by Kyle-Verhoog — APPROVED** [review]

**Review by Kyle-Verhoog — APPROVED** [review]

**Review by Kyle-Verhoog — APPROVED** [review]

**Review by majorgreys — APPROVED** [review]

**Comment by Yun-Kim** [comment]
@mergifyio backport 1.6

**Pull request merged**
